main/CTLIterator.ml

```diff
@@ -0,0 +1,549 @@
+  in
+  let rec addTerminationStmtReturn (block:block) = match block with
+    | A_empty l -> block 
+    | A_block (l, (A_return,a) , nextBlock) -> 
+      let nextBlock = A_block (l, (A_return, a), addTerminationStmtReturn nextBlock) in
+      A_block (nextId (), assignOne, nextBlock)
+    | A_block (l, stmt, nextBlock) -> A_block (l, stmt, addTerminationStmtReturn nextBlock)
+  in
+  let augmentedBody = A_block (-1, assignZero, addTerminationStmtReturn @@ addTerminationStmt mainFunction.funcBody) in
+  let v1 = snd (List.split (StringMap.bindings globalVariables)) in
+  let v2 = snd (List.split (StringMap.bindings mainFunction.funcVars)) in
+  let vars = terminationVar :: (List.append v1 v2) in (* add special terminationVar to list of variables *)
+  let var_to_apron v = Apron.Var.of_string v.varId in
+  let apron_vars = Array.map var_to_apron (Array.of_list vars) in
+  let env = Environment.make apron_vars [||] in
+  let program = {
+    environment = env;
+    variables = vars;
+    mainFunction = {
+      funcName = mainFunction.funcName;
+      funcTyp = mainFunction.funcTyp;
+      funcArgs = mainFunction.funcArgs;
+      funcVars = StringMap.add terminationVar.varId terminationVar mainFunction.funcVars; (* add termination var to function variables *)
+      funcBody = augmentedBody
+    };
+    globalBlock = globalBlock;
+    dummyPosition = dummyPosition
+  } in
+  let terminationProperty = AF (Atomic (AbstractSyntax.A_rbinary (AbstractSyntax.A_GREATER_EQUAL, (A_var terminationVar, dummyPosition), (A_const 1, dummyPosition)))) in
+  (program, terminationProperty)
+
+
+let prog_of_program (program:program) : prog = 
+  let funcMap = StringMap.add "main" program.mainFunction StringMap.empty in
+  let varMap = List.fold_left (fun map var -> StringMap.add var.varId var map) StringMap.empty program.variables in
+  (varMap, program.globalBlock, funcMap)
+
+(* bundle values into 'program' struct *)
+let program_of_prog (prog: AbstractSyntax.prog) (main: AbstractSyntax.StringMap.key) : program =
+  let (globalVariables, globalBlock, functions) = prog in
+  let mainFunction = StringMap.find main functions in
+  let v1 = snd (List.split (StringMap.bindings globalVariables)) in
+  let v2 = snd (List.split (StringMap.bindings mainFunction.funcVars)) in
+  let vars = List.append v1 v2 in
+  let var_to_apron v = Apron.Var.of_string v.varId in
+  let apron_vars = Array.map var_to_apron (Array.of_list vars) in
+  let env = Environment.make apron_vars [||] in
+  let dummyPosition = match mainFunction.funcBody with (* store some syntax possition for later use *)
+    | A_block (_, (_,pos), _) -> pos
+    | _ -> raise (Invalid_argument "empty main function")
+  in
+  {
+    environment = env;
+    variables = vars;
+    mainFunction = mainFunction;
+    globalBlock = globalBlock;
+    dummyPosition = dummyPosition
+  }
+
+
+module ForwardIterator (D:RANKING_FUNCTION) = struct
```

Comment:
I also thought about doing something like this. In the future, I could move the forward analysis to a dedicated functor and change the termination, guarantee and recurrence iterators accordingly.